### DISCUSSION TITLE
Is this expected behavior?

### DISCUSSION THREAD

**Discussion by ahron1** [invertase/react-native-firebase#5423] [Q&A] [answered]
Hi, 

I am seeing some strange behavior and hopefully someone can help me sort out what's wrong. Or if it is normal. 

Small note - during development, I use two mobile numbers for testing. Often, I signout one number and then sign in using the other. 

First I check if `auth.currentUser` is available, and if it is, I set some local states with user details and also with the FCM token. 

```
  const [phoneNumber, setPhoneNumber] = useState(null);
  const [fireBaseUid, setFireBaseUid] = useState(null);
  const [fcmToken, setFcmToken] = useState(null);


  useEffect(() => {
    if (auth().currentUser) {
      setFireBaseUid(auth().currentUser.uid);
      setPhoneNumber(auth().currentUser.phoneNumber);
      messaging()
        .getToken()
        .then(token => {
          setFcmToken(token);
        });
    } 
  }, [auth().currentUser]);
```

Then, if/when the states are updated, I send these values over to my server. 

```

  useEffect(() => {
    if (fireBaseUid && phoneNumber && fcmToken) {
      credUpdate({fireBaseUid, phoneNumber, fcmToken})(authDispatch);
    } 
  }, [fireBaseUid, phoneNumber, fcmToken]);
```

On the server, these values are written to a database. 
Oddly enough, it seems to record identical FCM Token values for different UIDs (corresponding to the different numbers). 
https://ibb.co/tDhzVRb 

Is this expected behavior? Does it mean the FCM token is tied only to the device and has nothing to do with the user account?

**Comment by mikehardy** [comment, marked as answer]
FCM is per app installation, under the covers tokens are provided by the "Installations" SDK module, which is the clue to know their uniqueness (but is not really surfaced anywhere, so is not that obvious)

When you log a user out, you should use these two methods back-to-back, waiting for promise resolution of course

https://rnfirebase.io/reference/messaging#deleteToken
https://rnfirebase.io/reference/messaging#getToken

**Reply by ahron1** [reply]
Ok, thanks. 

After posting I looked around a bit, and found this SO question with good answers. Linking it [here](https://stackoverflow.com/questions/43193215/firebase-cloud-messaging-handling-logout) in case any future readers might find it useful.

**Reply by mikehardy** [reply]
That SO is pretty old, lots of things have changed since then. I would just think "tokens are related to my app instance, if I log a user out some other user may log in meaning the tokens are no longer correct, so on logout I should delete the token, and on login I should get the token". It's no more complicated than that.

Also note that in react-native-firebase v12 there was a breaking change where tokens no longer have scopes, they are just simple method calls, no fuss.

**Reply by ahron1** [reply]
So when the previous user logs back in again, they get a new token, and the old (now deleted) token is basically gone forever, right?

**Reply by mikehardy** [reply]
If you delete a token, it is invalid immediately and forever.
When you perform these operations and how you handle the tokens and logout/login is up to the project and your business requirements

**Reply by ahron1** [reply]
Clear. Thanks a lot for explaining. 

Is there any place this stuff is documented in any detail?

**Reply by mikehardy** [reply]
Not really. Perhaps if it's unclear to new adopters of the library, you could hit the edit link at top right of the messaging getToken reference doc (and or in the usage doc if it exists there?) and add a note that tokens are associated with the app install, so if you have users login and logout on your app you should cycle the token in concert

**Reply by ahron1** [reply]
Yes, good idea. I will do that tomorrow, write up a small note and append it to the doc.

**Reply by ahron1** [reply]
> Not really. Perhaps if it's unclear to new adopters of the library, you could hit the edit link at top right of the messaging getToken reference doc (and or in the usage doc if it exists there?) and add a note that tokens are associated with the app install, so if you have users login and logout on your app you should cycle the token in concert

The usage page doesn't have anything much about tokens, I think the usage doc is more about the frontend side of things. The [server integration doc](https://rnfirebase.io/messaging/server-integration) on the other hand has a section on tokens and a good amount of relevant information (but nothing about what is discussed in this thread). So I added a paragraph to that doc instead and made a PR. What do you say? 

On a related note, the usage doc for Firebase Auth doesn't mention anything about how to get the auth token (by using the `getIdToken` method)  for authenticating with the backend, as mentioned in the [original Firebase docs ](https://firebase.google.com/docs/auth/web/manage-users#get_a_users_profile). So I am updating that too.
